qiling/utils.py

```diff
@@ -404,32 +404,37 @@ def select_os(ostype: QL_OS) -> QlClassInit['QlOs']:
 
     return partial(obj)
 
-def profile_setup(ostype: QL_OS, filename: Optional[str]):
+def profile_setup(ostype: QL_OS, config_src: Optional[Union[str, dict]]):
     # mcu uses a yaml-based config
     if ostype == QL_OS.MCU:
         import yaml
 
-        if filename:
-            with open(filename) as f:
+        if config_src:
+            with open(config_src) as f:
                 config = yaml.load(f, Loader=yaml.Loader)
         else:
             config = {}
 
     else:
-        qiling_home = Path(inspect.getfile(inspect.currentframe())).parent
-        os_profile = qiling_home / 'profiles' / f'{ostype.name.lower()}.ql'
-
-        profiles = [os_profile]
-
-        if filename:
-            profiles.append(filename)
-
         # patch 'getint' to convert integers of all bases
         int_converter = partial(int, base=0)
-
         config = ConfigParser(converters={'int': int_converter})
-        config.read(profiles)
 
+        # build config from dictionary, but not read from disk file
+        if isinstance(config_src, dict):
+            config.read_dict(config_src)
+
+        else:
+            qiling_home = Path(inspect.getfile(inspect.currentframe())).parent
+            os_profile = qiling_home / 'profiles' / f'{ostype.name.lower()}.ql'
+
+            profiles = [os_profile]
```

Comment:
I don't know whether this is intentional or not, but when a filename is specified, the user-specified profile file will be appeneded to the default profile (that is, there are defaults and user may provide specific overrides). The new addition doesn't support that, so if a user provides a dictionary rather than a filename, then there are no defaults - just the mapping on the specified dictionary.

If we want to retain the same behavior also for diciotnary-based profile, then we need to extract this line (i.e. `profiles = [os_profile]`) out.